widget/numerical_entry.go

```diff
@@ -26,13 +28,23 @@ func NewNumericalEntry() *NumericalEntry {
 //
 // Implements: fyne.Focusable
 func (e *NumericalEntry) TypedRune(r rune) {
+	if e.Entry.CursorColumn == 0 && e.Entry.CursorRow == 0 &&
+		len(e.Entry.Text) > 0 && e.Entry.Text[0] == '-' {
+		return
+	}
+
 	if r >= '0' && r <= '9' {
 		e.Entry.TypedRune(r)
 		return
 	}
 
 	if e.AllowFloat && (r == '.' || r == ',') {
 		e.Entry.TypedRune(r)
+		return
+	}
+
+	if r == '-' && e.AllowNegative && e.Entry.CursorRow == 0 && e.Entry.CursorColumn == 0 {
+		e.Entry.TypedRune(r)
 	}
```

Comment:
You can group the two new cases together to avoid some code duplication. They are more or less handling the same thing now. Something like this would be a bit cleaner I think (note that it is untested):

```go
if e.AllowNegative && e.Entry.CursorColumn == 0 && e.Entry.CursorRow == 0 {
  if len(e.Entry.Text) == 0 || len(e.Entry.Text) > 0 && e.Entry.Text[0] != '-' {
    e.Entry.OnTyped()
}
```

Reply:
I ended up needing three levels of ifs, but my changes do pass all tests.